File: mars/lib/parser.dart
import 'package:mars/mars.dart';
import 'package:petitparser/petitparser.dart';

class RobotsDefinition extends GrammarDefinition {
  @override
  Parser start() => ref0(range).trim() & ref0(robot).trim().star();

  Parser<Grid> range() => (ref0(number).trim() & ref0(number).trim())
      .map((values) => Grid(values[0], values[1]));

  Parser<Robot> initial() =>
      (ref0(number).trim() & ref0(number).trim() & ref0(direction).trim())
          .map((values) => Robot(values[0], values[1], values[2]));

  // TODO: This is clumsy
  Parser<Direction> direction() =>
      (char('N') | char('E') | char('S') | char('W')).map((ch) {
        switch (ch) {
          case 'N':
            return Direction.north;
          case 'E':
            return Direction.east;
          case 'S':
            return Direction.south;
          case 'W':
            return Direction.west;
          default:
            throw ArgumentError.value(ch);
        }
      });

  Parser<Robot> robot() =>
      (ref0(initial).trim() & ref0(instructions).trim()).map((values) {
        Robot robot = values[0];
        List<Instruction> instructions = values[1];
        robot.instructions = instructions;
        return robot;
      });

  Parser<List<Instruction>> instructions() => ref0(instruction).star();
  Parser<Instruction> instruction() =>
      (char('L') | char('R') | char('F')).map((ch) {
        switch (ch) {
          case 'L':
            return Instruction.left;
          case 'R':
            return Instruction.right;
          case 'F':
            return Instruction.forward;
          default:
            throw ArgumentError.value(ch);
        }
      });

  Parser<int> number() =>
      (digit().plus().flatten()).map((value) => int.parse(value));
}

File: redplanet/src/parser.rs
use crate::model::*;
use chumsky::prelude::*;

pub fn parser() -> impl Parser<char, Problem, Error = Simple<char>> {
    let num = text::int(10).map(|s: String| s.parse().unwrap()).padded();

    let grid = num.then(num).map(|(x, y)| Grid { x, y });
    let direction = choice((
        just("N").to(Direction::North),
        just("E").to(Direction::East),
        just("S").to(Direction::South),
        just("W").to(Direction::West),
    ))
    .padded();
    let instruction = choice((
        just("L").to(Instruction::Left),
        just("R").to(Instruction::Right),
        just("F").to(Instruction::Forward),
    ));
    let robot = num
        .then(num)
        .then(direction)
        .then(instruction.repeated().padded())
        .map(|(((x, y), d), instructions)| Robot {
            x,
            y,
            direction: d,
            is_lost: false,
            instructions,
        });
    let robots = robot.repeated();
    grid.then(robots)
        .map(|(grid, robots)| Problem { grid, robots })
}
